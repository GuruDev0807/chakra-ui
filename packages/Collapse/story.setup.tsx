import theme from "@chakra-ui/preset-base";
import {
  ColorMode,
  ColorModeProvider,
  createThemeContext,
  CSSReset,
} from "@chakra-ui/system";
import * as React from "react";
import { Global } from "@emotion/core";
import { PortalManager } from "@chakra-ui/portal";

const [ThemeProvider, useTheme] = createThemeContext(theme);

const setup = (story: () => any) => (
  <PortalManager>
    <ThemeProvider>
      <ColorModeProvider>
        <CSSReset />
        <ColorMode />
        <Global
          styles={{
            "*": {
              fontFamily: "system-ui",
            },
          }}
        />
        {story()}
      </ColorModeProvider>
    </ThemeProvider>
  </PortalManager>
);

export { ThemeProvider, useTheme };
export default setup;
