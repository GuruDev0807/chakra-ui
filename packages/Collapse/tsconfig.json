{
  "extends": "../../tsconfig.build.json",
  "include": ["src", "types"],
  "exclude": ["tests", "story"]
}
