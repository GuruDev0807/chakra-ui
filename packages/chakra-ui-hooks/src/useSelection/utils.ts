interface GetNextIndex {
  step?: number;
  currentIndex: number;
  itemsLength: number;
  loop: boolean;
}

export function getNextIndex({
  step = 1,
  currentIndex,
  itemsLength,
  loop,
}: GetNextIndex) {
  if (currentIndex === -1) {
    return step > 0 ? 0 : itemsLength - 1;
  }

  const nextIndex = currentIndex + step;
  if (nextIndex < 0) {
    return loop ? itemsLength - 1 : 0;
  }

  if (nextIndex >= itemsLength) {
    return loop ? 0 : itemsLength - 1;
  }

  return nextIndex;
}

export function getItemIndex<T>(items: T[], item: T) {
  if (item == null || items.length === 0) {
    return -1;
  }
  return items.indexOf(item);
}

export function getValue<T>(valueProp: T, valueState: T) {
  const isControlled = valueProp != null;
  const value = isControlled ? valueProp : valueState;
  return [value, isControlled] as [T, boolean];
}
