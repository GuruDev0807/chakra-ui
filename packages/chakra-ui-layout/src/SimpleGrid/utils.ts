import { SystemProps } from "@chakra-ui/system";
import { ResponsiveValue } from "styled-system";

// These helper fns are modified versions of the amazing rebass library
// https://github.com/rebassjs/rebass/blob/master/packages/layout/src/index.js

const px = (n: string | number) => (typeof n === "number" ? n + "px" : n);

export function widthToColumns(
  width: SystemProps["width"],
): SystemProps["gridTemplateColumns"] | null {
  if (Array.isArray(width)) {
    return width.map(widthToColumns as any);
  }

  if (
    width !== null &&
    typeof width === "object" &&
    Object.keys(width).length > 0
  ) {
    const acc: Record<string, string> = {};
    for (let key in width) {
      acc[key] = `repeat(auto-fit, minmax(${px(width[key])}, 1fr))`;
    }
    return acc;
  }

  if (width != null) {
    return `repeat(auto-fit, minmax(${px(width as any)}, 1fr))`;
  }

  return null;
}

export function countToColumns(
  count: ResponsiveValue<number> | undefined,
): SystemProps["gridTemplateColumns"] | null {
  if (Array.isArray(count)) {
    return count.map(countToColumns as any);
  }

  if (
    count !== null &&
    typeof count === "object" &&
    Object.keys(count).length > 0
  ) {
    const acc: Record<string, string> = {};
    for (let key in count) {
      acc[key] = `repeat(${count[key]}, 1fr)`;
    }
    return acc;
  }

  if (count != null) {
    return `repeat(${count}, 1fr)`;
  }

  return null;
}
