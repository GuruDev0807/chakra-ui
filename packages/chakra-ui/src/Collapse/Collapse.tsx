/** @jsx jsx */
import { jsx } from "@emotion/core";
import { forwardRef } from "react";
import AnimateHeight, { AnimateHeightProps } from "react-animate-height";
import { Box, BoxProps } from "@chakra-ui/layout";

type AnimateHeightOptions = Pick<
  AnimateHeightProps,
  | "animationStateClasses"
  | "applyInlineTransitions"
  | "delay"
  | "easing"
  | "style"
  | "children"
>;

interface CollapseOptions {
  /**
   * If `true`, the content will be visible
   */
  isOpen?: boolean;
  /**
   * If `true`, the opacity of the content will be animated
   */
  animateOpacity?: boolean;
  /**
   * The duration of the animation in `ms`
   */
  duration?: number;
  /**
   * The height you want the content in it's collapsed state. Set to `0` by default
   */
  startingHeight?: number | string;
  /**
   * The height you want the content in it's expanded state. Set to `auto` by default
   */
  endingHeight?: number | string;
  /**
   * The function to be called when the collapse animation starts. It provides the `newHeight` as an argument
   */
  onAnimationEnd?(props: { newHeight: number }): void;
  /**
   * The function to be called when the collapse animation ends. It provides the `newHeight` as an argument
   */
  onAnimationStart?(props: { newHeight: number }): void;
}

export type CollapseProps<P, T> = AnimateHeightOptions &
  CollapseOptions &
  BoxProps<P, T>;

const Collapse = forwardRef(function Collapse<P = {}, T = HTMLElement>(
  {
    isOpen,
    animateOpacity = true,
    onAnimationStart,
    onAnimationEnd,
    duration,
    easing = "ease",
    startingHeight = 0,
    endingHeight = "auto",
    ...rest
  }: CollapseProps<P, T>,
  ref: React.Ref<any>,
) {
  return (
    <AnimateHeight
      duration={duration}
      easing={easing}
      animateOpacity={animateOpacity}
      height={isOpen ? endingHeight : startingHeight}
      applyInlineTransitions={false}
      css={{
        transition:
          "height .2s ease,opacity .2s ease-in-out,transform .2s ease-in-out",
        "&.rah-animating--to-height-zero": {
          opacity: 0,
          transform: "translateY(-0.625rem)",
        },
      }}
      {...{ onAnimationStart, onAnimationEnd }}
    >
      <Box ref={ref} {...rest} />
    </AnimateHeight>
  );
}) as <P, T>(
  props: CollapseProps<P, T>,
) => React.ReactElement<CollapseProps<P, T>>;

export default Collapse;
