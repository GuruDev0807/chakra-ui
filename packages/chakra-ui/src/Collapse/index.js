/** @jsx jsx */
import { jsx } from "@emotion/core";
import { forwardRef } from "react";
import AnimateHeight from "react-animate-height";
import Box from "../Box";

const Collapse = forwardRef(
  (
    {
      isOpen,
      animateOpacity = true,
      onAnimationStart,
      onAnimationEnd,
      duration,
      easing = "ease",
      startingHeight = 0,
      endingHeight = "auto",
      ...rest
    },
    ref,
  ) => {
    return (
      <AnimateHeight
        duration={duration}
        easing={easing}
        animateOpacity={animateOpacity}
        height={isOpen ? endingHeight : startingHeight}
        applyInlineTransitions={false}
        css={{
          transition:
            "height .2s ease,opacity .2s ease-in-out,transform .2s ease-in-out",
          "&.rah-animating--to-height-zero": {
            opacity: 0,
            transform: "translateY(-0.625rem)",
          },
        }}
        {...{ onAnimationStart, onAnimationEnd }}
      >
        <Box ref={ref} {...rest} />
      </AnimateHeight>
    );
  },
);

Collapse.displayName = "Collapse";

export default Collapse;
