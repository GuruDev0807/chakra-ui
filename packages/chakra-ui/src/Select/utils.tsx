export function registerOption(items: any[], item: any) {
  if (items.length === 0) {
    return [item];
  }
  const index = items.findIndex(_item => _item.id === item.id);
  if (index >= 0) return items;

  const indexAfter = items.findIndex(
    _item =>
      !!(
        _item.ref.current &&
        item.ref.current &&
        item.ref.current.compareDocumentPosition(item.ref.current) &
          Node.DOCUMENT_POSITION_PRECEDING
      ),
  );

  if (indexAfter === -1) {
    return [...items, item];
  }

  return [...items.slice(0, indexAfter), item, ...items.slice(indexAfter)];
}
