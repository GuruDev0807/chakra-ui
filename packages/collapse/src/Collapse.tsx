import { chakra, PropsOf } from "@chakra-ui/system"
import {
  Transition,
  TransitionProps,
  TransitionStyles,
} from "@chakra-ui/transition"
import { ariaAttr, cx, mergeRefs, __DEV__ } from "@chakra-ui/utils"
import { useRect } from "@reach/rect"
import * as React from "react"
import {
  cloneElement,
  forwardRef,
  useState,
  Children,
  useRef,
  Ref,
} from "react"

export type CollapseProps = PropsOf<typeof chakra.div> & {
  isOpen?: boolean
  startingHeight?: number
  config?: TransitionProps["styles"]
  animateOpacity?: boolean
  transition?: string
}

export const Collapse = forwardRef(
  (props: CollapseProps, forwardedRef: React.Ref<any>) => {
    const {
      isOpen,
      children,
      config,
      startingHeight = 0,
      animateOpacity = true,
      className,
      style: htmlStyle,
      transition = "height 200ms ease, opacity 200ms ease, transform 200ms ease",
      ...rest
    } = props

    const [hidden, setHidden] = useState(true)

    type ChildElement = React.ReactElement<{ ref: Ref<any> }>

    let child = children

    if (typeof children === "string") {
      console.warn(
        `Warning: You're using a string directly inside <Collapse>. We recommend that you add an <div> tag as child of <Collapse>`,
      )
      child = <div>{children}</div> // fallback
    }

    const _child = Children.only(child) as ChildElement

    const ref = useRef<HTMLDivElement>(null)

    const rect = useRect(ref, true)
    const height = rect?.height ?? 0

    const styles: TransitionStyles = {
      init: {
        height: startingHeight,
        opacity: startingHeight ? 1 : 0,
      },
      entered: {
        height,
        opacity: 1,
        transform: "translateY(0)",
      },
      exiting: {
        height: startingHeight,
        opacity: startingHeight ? 1 : 0,
        transform: "translateY(-0.5rem)",
      },
    }

    return (
      <Transition
        in={isOpen}
        styles={config || styles}
        onEntered={() => setHidden(false)}
        onExited={() => setHidden(true)}
        timeout={{ enter: 50, exit: 200 }}
        transition={transition}
        unmountOnExit={false}
      >
        {styles => (
          <chakra.div
            ref={forwardedRef}
            className={cx("chakra-collapse", className)}
            aria-hidden={ariaAttr(hidden)}
            {...rest}
            style={{
              ...styles,
              overflow: "hidden",
              opacity: animateOpacity ? styles.opacity : 1,
              willChange: "height, opacity",
              ...htmlStyle,
            }}
          >
            {cloneElement(_child, {
              ref: mergeRefs(ref, _child.props.ref),
            })}
          </chakra.div>
        )}
      </Transition>
    )
  },
)

if (__DEV__) {
  Collapse.displayName = "Collapse"
}
