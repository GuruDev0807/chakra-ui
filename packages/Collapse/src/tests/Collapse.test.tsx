import "@testing-library/jest-dom/extend-expect"
import React from "react"
import { render } from "../../../../test/utils"
import { Collapse } from "../Collapse"

test("Collapse renders correctly", () => {
  const utils = render(<Collapse>sfsjdfkbfkjdfdjksf</Collapse>)
  expect(utils.asFragment()).toMatchSnapshot()
})
