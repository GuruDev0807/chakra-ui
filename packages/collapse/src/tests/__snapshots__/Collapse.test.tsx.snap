// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="chakra-collapse"
    style="height: 0px; opacity: 0; transition: height 200ms ease, opacity 200ms ease, transform 200ms ease; overflow: hidden; will-change: height, opacity;"
  >
    <div>
      sfsjdfkbfkjdfdjksf
    </div>
  </div>
</DocumentFragment>
`;
